providers/NMSReflectionProvider/src/main/java/net/ess3/nms/refl/providers/ReflPersistentDataProvider.java

```diff
@@ -0,0 +1,125 @@
+
+    public ReflPersistentDataProvider(Plugin plugin) {
+        this.namespace = plugin.getName().toLowerCase(Locale.ROOT);
+
+        MethodHandle itemHandleGetterHandle = null;
+        MethodHandle getTagHandle = null;
+        MethodHandle tagSetterHandle = null;
+        MethodHandle newCompoundHandle = null;
+        MethodHandle getCompoundHandle = null;
+        MethodHandle setCompoundHandle = null;
+        MethodHandle setStringHandle = null;
+        MethodHandle removeHandle = null;
+        MethodHandle getStringHandle = null;
+        try {
+            final MethodHandles.Lookup lookup = MethodHandles.lookup();
+            final Field handleGetter = ReflUtil.getOBCClass("inventory.CraftItemStack").getDeclaredField("handle");
+            handleGetter.setAccessible(true);
+            itemHandleGetterHandle = lookup.unreflectGetter(handleGetter);
+            final Field tagSetter = ReflUtil.getNMSClass("ItemStack").getDeclaredField("tag");
+            tagSetter.setAccessible(true);
+            tagSetterHandle = lookup.unreflectSetter(tagSetter);
+            getTagHandle = lookup.findVirtual(ReflUtil.getNMSClass("ItemStack"), "getTag", MethodType.methodType(ReflUtil.getNMSClass("NBTTagCompound")));
+            newCompoundHandle = lookup.findConstructor(ReflUtil.getNMSClass("NBTTagCompound"), MethodType.methodType(void.class));
+            getCompoundHandle = lookup.findVirtual(ReflUtil.getNMSClass("NBTTagCompound"), "getCompound", MethodType.methodType(ReflUtil.getNMSClass("NBTTagCompound"), String.class));
+            setCompoundHandle = lookup.findVirtual(ReflUtil.getNMSClass("NBTTagCompound"), "set", MethodType.methodType(void.class, String.class, ReflUtil.getNMSClass("NBTBase")));
+            setStringHandle = lookup.findVirtual(ReflUtil.getNMSClass("NBTTagCompound"), "setString", MethodType.methodType(void.class, String.class, String.class));
+            removeHandle = lookup.findVirtual(ReflUtil.getNMSClass("NBTTagCompound"), "remove", MethodType.methodType(void.class, String.class));
+            getStringHandle = lookup.findVirtual(ReflUtil.getNMSClass("NBTTagCompound"), "getString", MethodType.methodType(String.class, String.class));
+        } catch (NoSuchFieldException | IllegalAccessException | NoSuchMethodException e) {
+            e.printStackTrace();
+        }
+        this.itemHandleGetterHandle = itemHandleGetterHandle;
+        this.getTagHandle = getTagHandle;
+        this.tagSetterHandle = tagSetterHandle;
+        this.newCompoundHandle = newCompoundHandle;
+        this.getCompoundHandle = getCompoundHandle;
+        this.setCompoundHandle = setCompoundHandle;
+        this.setStringHandle = setStringHandle;
+        this.removeHandle = removeHandle;
+        this.getStringHandle = getStringHandle;
+    }
+
+    private String getPersistentString(ItemStack itemStack, String key) throws Throwable {
+        final Object nmsItem = itemHandleGetterHandle.invoke(itemStack);
```

Comment:
You could _probably_ call `#invokeExact` here, for the little gains that may provide.

Reply:
When MethodHandles mean exact, they mean exact. It will not even perform downcasting to arguments/instances passed into it.